### PR TITLE
ODSC-46634/utilize oci UploadManager to upload model artifacts

### PR DESCRIPTION
**Pull request by mingkang111**
# Description
Utilize the [UploadManager](https://docs.oracle.com/en-us/iaas/tools/python/2.104.1/api/upload_manager.html) in the [ArtifactUploader](https://github.com/oracle/accelerated-data-science/blob/main/ads/model/artifact_uploader.py) to upload model artifacts to the object storage.
https://jira.oci.oraclecorp.com/browse/ODSC-46634

For details see this [doc](https://confluence.oci.oraclecorp.com/display/~mingkang/Upload+Model+Artifacts).

# User experience
* Adds argument `PARALLEL_PROCESS_COUNT` in the following public APIs. 
```
ads.model.GenericModel.save
ads.model.GenericModel.upload_artifact
```
* After this change, the time used in `save()` for large model artifacts was reduced by ~40%.
**Test with model artifact ~=2.3G
Screenshot 2023-09-05 at 9 58 43 AM

Screenshot 2023-09-05 at 12 32 52 PM

##  Pre-commit
Screenshot 2023-09-05 at 9 57 19 AM

### CODE REVIEW

File: ads/common/utils.py [hunk 1]
```diff
@@ -1599,3 +1603,35 @@ def is_path_exists(uri: str, auth: Optional[Dict] = None) -> bool:
     if fsspec.filesystem(path_scheme, **storage_options).exists(uri):
         return True
     return False
+
+
+def parse_os_uri(uri: str):
```
Comment (mrDzurb) [inline ads/common/utils.py:1608]: Wouldn't it be better to use   - ```ObjectStorageDetails``` instead?

File: ads/model/artifact_uploader.py [hunk 2]
```diff
@@ -104,6 +110,39 @@ def _upload(self):
 
 
 class LargeArtifactUploader(ArtifactUploader):
+    """
+    The class helper to upload large model artifacts.
+
+    Attributes
```
Comment (mrDzurb) [inline ads/model/artifact_uploader.py:116]: Nice!

File: ads/model/artifact_uploader.py [hunk 3]
```diff
@@ -150,36 +189,49 @@ def __init__(
...
             raise FileExistsError(
-                f"The `{self.bucket_uri}` exists. Please use a new file name or "
+                f"The bucket_uri=`{self.bucket_uri}` exists. Please use a new file name or "
                 "set `overwrite_existing_artifact` to `True` if you wish to overwrite."
             )
+
+        bucket_name, namespace_name, object_name = utils.parse_os_uri(bucket_uri)
+        logger.debug(f"{bucket_name=}, {namespace_name=}, {object_name=}")
+        try:
+            response = self.upload_manager.upload_file(
```
Comment (mrDzurb) [inline ads/model/artifact_uploader.py:219]: Does it show a progress bar?

Comment (mingkang111) [reply ads/model/artifact_uploader.py]: Improved to show progress bar.

File: ads/common/utils.py (hunk 1 above)
Comment (mingkang111) [reply ads/common/utils.py]: Changed to use `ObjectStorageDetails` instead.

**Review by mrDzurb — APPROVED** [review]

**Review by mrDzurb — APPROVED** [review]

**Review by z7ye — APPROVED** [review]

**Pull request merged**